pkg/policies/action/action.go

```diff
@@ -0,0 +1,757 @@
+
+	topLangBytes := 0
+	topLang := ""
+
+	var significantLanguages []string
+
+	for l, b := range langs {
+		totalBytes += b
+		if topLang == "" || topLangBytes < b {
+			topLang = l
+			topLangBytes = b
+		}
+		if b > 3000 {
+			significantLanguages = append(significantLanguages, l)
+		}
+	}
+
+	significantLanguages = append(significantLanguages, topLang)
+
+	for _, w := range want {
+		for _, s := range significantLanguages {
+			if strings.EqualFold(s, w) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// Len returns number of rules in s
+func (s sortableRules) Len() int {
+	return len(s)
+}
+
+// Less returns i < j, determined by Priority and Method so that priority tiers
+// are honored and allow/require is evaluated before deny.
+func (s sortableRules) Less(i, j int) bool {
+	if s[i].priorityInt < s[j].priorityInt {
+		return true
+	}
+	if s[i].Method != "deny" {
+		return true
+	}
+	return false
+}
+
+// Swap swaps the Rules at indices i and j
+func (s sortableRules) Swap(i, j int) {
+	hold := s[i]
+	s[i] = s[j]
+	s[j] = hold
+}
+
+// githubGetLatestCommitHash gets the latest commit hash for a repo's default
+// branch using the GitHub API.
+// Docs: https://docs.github.com/en/rest/commits/commits#list-commits
+func githubGetLatestCommitHash(ctx context.Context, c *github.Client, owner, repo string) (string, error) {
+	commits, _, err := c.Repositories.ListCommits(ctx, owner, repo, &github.CommitsListOptions{})
+	if err != nil {
+		return "", err
+	}
+	if len(commits) > 0 && commits[0].SHA != nil {
+		return *commits[0].SHA, nil
+	}
+	return "", fmt.Errorf("repo has no commits: %w", err)
+}
+
+// githubListWorkflowRunsByFilename returns workflow runs for a repo by
+// workflow filename.
+// Docs:
+// https://docs.github.com/en/rest/actions/workflow-runs#list-workflow-runs
+func githubListWorkflowRunsByFilename(ctx context.Context, c *github.Client, owner, repo string, workflowFilename string) ([]*github.WorkflowRun, error) {
+	runs, _, err := c.Actions.ListWorkflowRunsByFileName(ctx, owner, repo, workflowFilename, &github.ListWorkflowRunsOptions{
+		Event: "push",
+	})
+	return runs.WorkflowRuns, err
+}
+
+// githubListLanguages uses the GitHub API to list languages.
+// Docs: https://docs.github.com/en/rest/repos/repos#list-repository-languages
+func githubListLanguages(ctx context.Context, c *github.Client, owner, repo string) (map[string]int, error) {
+	l, _, err := c.Repositories.ListLanguages(ctx, owner, repo)
+	return l, err
+}
+
+// githubListWorkflows returns workflows for a repo. If on is specified, will
+// filter to workflows with all trigger events listed in on.
+// Docs: https://docs.github.com/en/rest/repos/contents#get-repository-content
+func githubListWorkflows(ctx context.Context, c *github.Client, owner, repo string, on []string) ([]*workflowMetadata, error) {
+	_, workflowDirContents, resp, err := c.Repositories.GetContents(ctx, owner, repo, ".github/workflows", &github.RepositoryContentGetOptions{})
```

Comment:
404s are not cacheable, so we like to walk the tree from root. The config does it with pkg/config/contents.go. That could be made public in the config package. If it's too tricky for now, just add a FIXME/TODO